File: libraries/libym_logic/tests/PyUnit/ExprTest.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

import YmLogic
import unittest

class ExprTest(unittest.TestCase) :

    # 初期化
    def setUp(self) :
        pass

    # 終了処理
    def tearDown(self) :
        pass

    # 空のテスト
    def test_empty(self) :
        pass

    # 空のコンストラクタのテスト
    def test_constructor1(self) :
        expr = YmLogic.Expr()

    # 定数0ノードの検査
    def test_const0(self) :
        expr = YmLogic.Expr.make_zero()

        self.assertTrue( expr.is_zero() )
        self.assertFalse( expr.is_one() )
        self.assertTrue( expr.is_constant() )
        self.assertFalse( expr.is_posiliteral() )
        self.assertFalse( expr.is_negaliteral() )
        self.assertFalse( expr.is_literal() )
        self.assertFalse( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertFalse( expr.is_op() )
        self.assertEqual( 0, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertFalse( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 0, expr.litnum() )
        self.assertEqual( 0, expr.input_size() )
        self.assertEqual( 0, expr.sop_cubenum() )
        self.assertEqual( 0, expr.sop_litnum() )

    # 定数1ノードの検査
    def test_const1(self) :
        expr = YmLogic.Expr.make_one()

        self.assertFalse( expr.is_zero() )
        self.assertTrue( expr.is_one() )
        self.assertTrue( expr.is_constant() )
        self.assertFalse( expr.is_posiliteral() )
        self.assertFalse( expr.is_negaliteral() )
        self.assertFalse( expr.is_literal() )
        self.assertFalse( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertFalse( expr.is_op() )
        self.assertEqual( 0, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertFalse( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 0, expr.litnum() )
        self.assertEqual( 0, expr.input_size() )
        self.assertEqual( 0, expr.sop_cubenum() )
        self.assertEqual( 0, expr.sop_litnum() )

    # リテラルノードの検査(1)
    def test_literal1(self) :
        vnum = 1
        varid = YmLogic.VarId(vnum)
        inv = False
        expr = YmLogic.Expr.make_literal(varid, inv)

        self.assertFalse( expr.is_zero() )
        self.assertFalse( expr.is_one() )
        self.assertFalse( expr.is_constant() )
        self.assertTrue( expr.is_posiliteral() )
        self.assertFalse( expr.is_negaliteral() )
        self.assertTrue( expr.is_literal() )
        self.assertEqual( varid, expr.varid() )
        self.assertFalse( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertFalse( expr.is_op() )
        self.assertEqual( 0, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertFalse( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 1, expr.litnum() )
        self.assertEqual( vnum + 1, expr.input_size() )
        self.assertEqual( 1, expr.sop_cubenum() )
        self.assertEqual( 1, expr.sop_litnum() )

    # リテラルノードの検査(2)
    def test_literal2(self) :
        vnum = 2
        varid = YmLogic.VarId(vnum)
        inv = True
        expr = YmLogic.Expr.make_literal(varid, inv)

        self.assertFalse( expr.is_zero() )
        self.assertFalse( expr.is_one() )
        self.assertFalse( expr.is_constant() )
        self.assertFalse( expr.is_posiliteral() )
        self.assertTrue( expr.is_negaliteral() )
        self.assertTrue( expr.is_literal() )
        self.assertEqual( varid, expr.varid() )
        self.assertFalse( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertFalse( expr.is_op() )
        self.assertEqual( 0, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertFalse( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 1, expr.litnum() )
        self.assertEqual( vnum + 1, expr.input_size() )
        self.assertEqual( 1, expr.sop_cubenum() )
        self.assertEqual( 1, expr.sop_litnum() )

    # リテラルノードの検査(3)
    def test_literal3(self) :
        vnum = 100
        varid = YmLogic.VarId(vnum)
        expr = YmLogic.Expr.make_posiliteral(varid)

        self.assertFalse( expr.is_zero() )
        self.assertFalse( expr.is_one() )
        self.assertFalse( expr.is_constant() )
        self.assertTrue( expr.is_posiliteral() )
        self.assertFalse( expr.is_negaliteral() )
        self.assertTrue( expr.is_literal() )
        self.assertEqual( varid, expr.varid() )
        self.assertFalse( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertFalse( expr.is_op() )
        self.assertEqual( 0, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertFalse( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 1, expr.litnum() )
        self.assertEqual( vnum + 1, expr.input_size() )
        self.assertEqual( 1, expr.sop_cubenum() )
        self.assertEqual( 1, expr.sop_litnum() )

    # リテラルノードの検査(4)
    def test_literal4(self) :
        vnum = 102
        varid = YmLogic.VarId(vnum)
        expr = YmLogic.Expr.make_negaliteral(varid)

        self.assertFalse( expr.is_zero() )
        self.assertFalse( expr.is_one() )
        self.assertFalse( expr.is_constant() )
        self.assertFalse( expr.is_posiliteral() )
        self.assertTrue( expr.is_negaliteral() )
        self.assertTrue( expr.is_literal() )
        self.assertEqual( varid, expr.varid() )
        self.assertFalse( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertFalse( expr.is_op() )
        self.assertEqual( 0, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertFalse( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 1, expr.litnum() )
        self.assertEqual( vnum + 1, expr.input_size() )
        self.assertEqual( 1, expr.sop_cubenum() )
        self.assertEqual( 1, expr.sop_litnum() )

    # make_and の検査(1)
    def test_make_and1(self) :
        varid0 = YmLogic.VarId(0)
        varid1 = YmLogic.VarId(1)
        varid2 = YmLogic.VarId(2)

        lit0 = YmLogic.Literal(varid0, False)
        lit1 = YmLogic.Literal(varid1, False)
        lit2 = YmLogic.Literal(varid2, False)

        expr0 = YmLogic.Expr.make_literal(lit0)
        expr1 = YmLogic.Expr.make_literal(lit1)
        expr2 = YmLogic.Expr.make_literal(lit2)
        chd_list = [ expr0, expr1, expr2 ]
        expr = YmLogic.Expr.make_and(chd_list)

        self.assertFalse( expr.is_zero() )
        self.assertFalse( expr.is_one() )
        self.assertFalse( expr.is_constant() )
        self.assertFalse( expr.is_posiliteral() )
        self.assertFalse( expr.is_negaliteral() )
        self.assertFalse( expr.is_literal() )
        self.assertTrue( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertTrue( expr.is_op() )
        self.assertEqual( 3, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertTrue( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 3, expr.litnum() )
        self.assertEqual( 3, expr.input_size() )
        self.assertEqual( 1, expr.sop_cubenum() )
        self.assertEqual( 3, expr.sop_litnum() )
        self.assertTrue( expr_equal(expr0, expr.child(0)) )
        self.assertTrue( expr_equal(expr1, expr.child(1)) )
        self.assertTrue( expr_equal(expr2, expr.child(2)) )

    # make_and の検査(2)
    def test_make_and1(self) :
        varid0 = YmLogic.VarId(0)
        varid1 = YmLogic.VarId(1)
        varid2 = YmLogic.VarId(2)

        lit0 = YmLogic.Literal(varid0, False)
        lit1 = YmLogic.Literal(varid1, False)
        lit2 = YmLogic.Literal(varid2, False)

        expr0 = YmLogic.Expr.make_literal(lit0)
        expr1 = YmLogic.Expr.make_literal(lit1)
        expr2 = YmLogic.Expr.make_literal(lit2)
        chd_list = [ expr0, expr1, expr2 ]
        expr = YmLogic.Expr.make_and(chd_list)

        self.assertFalse( expr.is_zero() )
        self.assertFalse( expr.is_one() )
        self.assertFalse( expr.is_constant() )
        self.assertFalse( expr.is_posiliteral() )
        self.assertFalse( expr.is_negaliteral() )
        self.assertFalse( expr.is_literal() )
        self.assertTrue( expr.is_and() )
        self.assertFalse( expr.is_or() )
        self.assertFalse( expr.is_xor() )
        self.assertTrue( expr.is_op() )
        self.assertEqual( 3, expr.child_num() )
        self.assertTrue( expr.is_simple() )
        self.assertTrue( expr.is_simple_and() )
        self.assertFalse( expr.is_simple_or() )
        self.assertFalse( expr.is_simple_xor() )
        self.assertTrue( expr.is_sop() )
        self.assertEqual( 3, expr.litnum() )
        self.assertEqual( 3, expr.input_size() )
        self.assertEqual( 1, expr.sop_cubenum() )
        self.assertEqual( 3, expr.sop_litnum() )
        self.assertTrue( expr_equal(expr0, expr.child(0)) )
        self.assertTrue( expr_equal(expr1, expr.child(1)) )
        self.assertTrue( expr_equal(expr2, expr.child(2)) )

def expr_equal(left, right) :
    if left.is_zero() :
        return right.is_zero()

    if left.is_one() :
        return right.is_one()

    if left.is_posiliteral() :
        return right.is_posiliteral() and lit_equal(left, right)

    if left.is_negaliteral() :
        return right.is_negaliteral() and lit_equal(left, right)

    if left.is_and() :
        return right.is_and() and op_equal(left, right)

    if left.is_or() :
        return right.is_or() and op_equal(left, right)

    if left.is_xor() :
        return right.is_xor() and op_equal(left, right)

    return False

def lit_equal(left, right) :
    return left.varid() == right.varid()

def opt_equal(left, right) :
    if left.child_num() != right.child_num() :
        return False

    for i in range(left.child_num()) :
        if not expr_equal(left.child(i), right.child(i)) :
            return False

    return True

# ExprTest のテストスイート
suite = unittest.TestLoader().loadTestsFromTestCase(ExprTest)


if __name__ == '__main__' :
    unittest.TextTestRunner(verbosity=4).run(suite)

File: libraries/libym_logic/tests/CppUTest/expr/ExprTest.cc

/// @file ExprTest.cc
/// @brief Expr のテストプログラム
/// @author Yusuke Matsunaga (松永 裕介)
///
/// Copyright (C) 2013 Yusuke Matsunaga
/// All rights reserved.


#include "CppUTest/TestHarness.h"

#include "YmLogic/Expr.h"

using namespace nsYm;


TEST_GROUP(libym_logic_ExprTestGroup)
{
  TEST_SETUP() {
  }

  TEST_TEARDOWN() {
    Expr::clear_memory();
  }

};


// 空のテスト
TEST(libym_logic_ExprTestGroup, empty)
{
}

// 定数0ノードの検査
TEST(libym_logic_ExprTestGroup, const0)
{
  Expr expr = Expr::make_zero();
  CHECK( expr.is_zero() );
  CHECK( !expr.is_one() );
  CHECK( expr.is_constant() );
  CHECK( !expr.is_posiliteral() );
  CHECK( !expr.is_negaliteral() );
  CHECK( !expr.is_literal() );
  CHECK( !expr.is_and() );
  CHECK( !expr.is_or() );
  CHECK( !expr.is_xor() );
  CHECK( !expr.is_op() );
  LONGS_EQUAL( 0, expr.child_num() );
  CHECK( expr.is_simple() );
  CHECK( !expr.is_simple_and() );
  CHECK( !expr.is_simple_or() );
  CHECK( !expr.is_simple_xor() );
  CHECK( expr.is_sop() );
  LONGS_EQUAL( 0, expr.litnum() );
  LONGS_EQUAL( 0, expr.input_size() );
  LONGS_EQUAL( 0, expr.sop_cubenum() );
  LONGS_EQUAL( 0, expr.sop_litnum() );
}

// 定数1ノードの検査
TEST(libym_logic_ExprTestGroup, const1)
{
  Expr expr = Expr::make_one();
  CHECK( !expr.is_zero() );
  CHECK( expr.is_one() );
  CHECK( expr.is_constant() );
  CHECK( !expr.is_posiliteral() );
  CHECK( !expr.is_negaliteral() );
  CHECK( !expr.is_literal() );
  CHECK( !expr.is_and() );
  CHECK( !expr.is_or() );
  CHECK( !expr.is_xor() );
  CHECK( !expr.is_op() );
  LONGS_EQUAL( 0, expr.child_num() );
  CHECK( expr.is_simple() );
  CHECK( !expr.is_simple_and() );
  CHECK( !expr.is_simple_or() );
  CHECK( !expr.is_simple_xor() );
  CHECK( expr.is_sop() );
  LONGS_EQUAL( 0, expr.litnum() );
  LONGS_EQUAL( 0, expr.input_size() );
  LONGS_EQUAL( 0, expr.sop_cubenum() );
  LONGS_EQUAL( 0, expr.sop_litnum() );
}

// リテラルノードの検査(1)
TEST(libym_logic_ExprTestGroup, literal1)
{
  const ymuint vnum = 1;
  const VarId vid(vnum);
  const bool inv = false;
  Expr expr = Expr::make_literal(vid, inv);

  CHECK( !expr.is_zero() );
  CHECK( !expr.is_one() );
  CHECK( !expr.is_constant() );
  CHECK( expr.is_posiliteral() );
  CHECK( !expr.is_negaliteral() );
  CHECK( expr.is_literal() );
  CHECK( vid == expr.varid() );
  CHECK( !expr.is_and() );
  CHECK( !expr.is_or() );
  CHECK( !expr.is_xor() );
  CHECK( !expr.is_op() );
  LONGS_EQUAL( 0, expr.child_num() );
  CHECK( expr.is_simple() );
  CHECK( !expr.is_simple_and() );
  CHECK( !expr.is_simple_or() );
  CHECK( !expr.is_simple_xor() );
  CHECK( expr.is_sop() );
  LONGS_EQUAL( 1, expr.litnum() );
  LONGS_EQUAL( vnum + 1, expr.input_size() );
  LONGS_EQUAL( 1, expr.sop_cubenum() );
  LONGS_EQUAL( 1, expr.sop_litnum() );
}

// リテラルノードの検査(2)
TEST(libym_logic_ExprTestGroup, literal2)
{
  const ymuint vnum = 2;
  const VarId vid(vnum);
  const bool inv = false;
  const Literal lit(vid, inv);
  Expr expr = Expr::make_literal(lit);

  CHECK( !expr.is_zero() );
  CHECK( !expr.is_one() );
  CHECK( !expr.is_constant() );
  CHECK( expr.is_posiliteral() );
  CHECK( !expr.is_negaliteral() );
  CHECK( expr.is_literal() );
  CHECK( vid == expr.varid() );
  CHECK( !expr.is_and() );
  CHECK( !expr.is_or() );
  CHECK( !expr.is_xor() );
  CHECK( !expr.is_op() );
  LONGS_EQUAL( 0, expr.child_num() );
  CHECK( expr.is_simple() );
  CHECK( !expr.is_simple_and() );
  CHECK( !expr.is_simple_or() );
  CHECK( !expr.is_simple_xor() );
  CHECK( expr.is_sop() );
  LONGS_EQUAL( 1, expr.litnum() );
  LONGS_EQUAL( vnum + 1, expr.input_size() );
  LONGS_EQUAL( 1, expr.sop_cubenum() );
  LONGS_EQUAL( 1, expr.sop_litnum() );
}

// リテラルノードの検査(3)
TEST(libym_logic_ExprTestGroup, literal3)
{
  const ymuint vnum = 3;
  const VarId vid(vnum);
  Expr expr = Expr::make_posiliteral(vid);

  CHECK( !expr.is_zero() );
  CHECK( !expr.is_one() );
  CHECK( !expr.is_constant() );
  CHECK( expr.is_posiliteral() );
  CHECK( !expr.is_negaliteral() );
  CHECK( expr.is_literal() );
  CHECK( vid == expr.varid() );
  CHECK( !expr.is_and() );
  CHECK( !expr.is_or() );
  CHECK( !expr.is_xor() );
  CHECK( !expr.is_op() );
  LONGS_EQUAL( 0, expr.child_num() );
  CHECK( expr.is_simple() );
  CHECK( !expr.is_simple_and() );
  CHECK( !expr.is_simple_or() );
  CHECK( !expr.is_simple_xor() );
  CHECK( expr.is_sop() );
  LONGS_EQUAL( 1, expr.litnum() );
  LONGS_EQUAL( vnum + 1, expr.input_size() );
  LONGS_EQUAL( 1, expr.sop_cubenum() );
  LONGS_EQUAL( 1, expr.sop_litnum() );
}

// リテラルノードの検査(4)
TEST(libym_logic_ExprTestGroup, literal4)
{
  const ymuint vnum = 4;
  const VarId vid(vnum);
  Expr expr = Expr::make_negaliteral(vid);

  CHECK( !expr.is_zero() );
  CHECK( !expr.is_one() );
  CHECK( !expr.is_constant() );
  CHECK( !expr.is_posiliteral() );
  CHECK( expr.is_negaliteral() );
  CHECK( expr.is_literal() );
  CHECK( vid == expr.varid() );
  CHECK( !expr.is_and() );
  CHECK( !expr.is_or() );
  CHECK( !expr.is_xor() );
  CHECK( !expr.is_op() );
  LONGS_EQUAL( 0, expr.child_num() );
  CHECK( expr.is_simple() );
  CHECK( !expr.is_simple_and() );
  CHECK( !expr.is_simple_or() );
  CHECK( !expr.is_simple_xor() );
  CHECK( expr.is_sop() );
  LONGS_EQUAL( 1, expr.litnum() );
  LONGS_EQUAL( vnum + 1, expr.input_size() );
  LONGS_EQUAL( 1, expr.sop_cubenum() );
  LONGS_EQUAL( 1, expr.sop_litnum() );
}

// make_andの検査(1)
TEST(libym_logic_ExprTestGroup, make_and1)
{
  Literal lit0(VarId(0), false);
  Expr expr0 = Expr::make_literal(lit0);

  Literal lit1(VarId(1), true);
  Expr expr1 = Expr::make_literal(lit1);

  Literal lit2(VarId(2), true);
  Expr expr2 = Expr::make_literal(lit2);

  ExprVector child_list(3);
  child_list[0] = expr0;
  child_list[1] = expr1;
  child_list[2] = expr2;

  Expr expr = Expr::make_and(child_list);

  CHECK( !expr.is_zero() );
  CHECK( !expr.is_one() );
  CHECK( !expr.is_constant() );
  CHECK( !expr.is_posiliteral() );
  CHECK( !expr.is_negaliteral() );
  CHECK( !expr.is_literal() );
  CHECK( expr.is_and() );
  CHECK( !expr.is_or() );
  CHECK( !expr.is_xor() );
  CHECK( expr.is_op() );
  LONGS_EQUAL( child_list.size(), expr.child_num() );
  CHECK( expr.is_simple() );
  CHECK( expr.is_simple_and() );
  CHECK( !expr.is_simple_or() );
  CHECK( !expr.is_simple_xor() );
  CHECK( expr.is_sop() );
  LONGS_EQUAL( child_list.size(), expr.litnum() );
  LONGS_EQUAL( 3, expr.input_size() );
  LONGS_EQUAL( 1, expr.sop_cubenum() );
  LONGS_EQUAL( 3, expr.sop_litnum() );
}
